### ISSUE TITLE
Feature detection `typeof localStorage` trips up permissions, requires `--location` to even start

### ISSUE THREAD

**Issue by mitranim** [denoland/deno#10825]
I'm converting an isomorphic app where 99% of the code runs in both Node/Deno and browsers. It uses checks like `typeof X !== 'undefined' && isObj(X)` to detect various browser APIs and avoid using them when unavailable.

Deno trips up on either of the following, and requires `--location` to even start!

```js
typeof localStorage !== 'undefined'
typeof localStorage !== 'undefined' && isObj(localStorage)
```

This code also runs in browsers. Using Deno-specific checks (if any are available) would require it to feature-detect the `Deno` global first!

Browser globals may also be missing in browsers. The checks aren't _just_ for Deno. As one example, a few years ago, I've dealt with `localStorage` either missing or throwing exceptions on access in iOS Safari in private browsing mode.

The docs make it sound like `--location` is sort of a crutch that should be avoided. Shouldn't Deno allow a cleaner solution here?

**Comment by kitsonk** [comment]
ping @crowlKats probably need to find a way not to do the permission check on the `localStorage` getter, but do it another way.

**Comment by crowlKats** [comment]
This has been discussed many times in the discord server, there is no solution to do this. @nayeemrmn

**Comment by kitsonk** [comment]
There needs to be a way.  You can't have a permission error when feature detecting. There is always a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) which would return a `"object"` but not actually access the local storage (there could be other traps that could be useful as well).  The `getPrototypeOf` trap can be used for `instanceof` for example.

**Comment by lucacasonato** [comment]
We could also consider just not having the global at all if `--location` is not defined. That would be more like what we do for `--unstable` things.

**Comment by crowlKats** [comment]
I think that will be a cause for confusion

**Comment by ry** [comment]
The situation @mitranim has described is very real and problematic.

Luca's solution works. We'd have to make up for the confusion with better documentation.

Alternatively we can define a default location. But those discussions have gone nowhere, so we should default to having `localStorage` be undefined.

**Comment by kitsonk** [comment]
> I think that will be a cause for confusion

The current situation is just as confusing. I think having local storage `undefined` without a `--location` is a good solution and easily document.

I would also personally be in favour of an opaque origin that serves as the default if no `--location` is provided.

**Comment by nayeemrmn** [comment]
I've mostly come around to just having it be undefined without `--location`. But the argument was that people should see an error message and just supply a `--location` rather than the possibility of silently going down a code path which assumes it's not supported. Supplying a `--location` is easy. On the other hand, I no longer like that the traditional feature-testing pattern is broken...

> The docs make it sound like `--location` is sort of a crutch that should be avoided.

It used to be before `localStorage` was added... the list of acceptable use cases at the bottom could be updated I suppose.

**Comment by kt3k** [comment]
closed by #12548

Now you can access `localStorage` without specifying `--location`. ref: https://deno.land/manual@v1.16.4/runtime/web_storage_api
